powerstore/resource_snapshot.go

```diff
@@ -0,0 +1,302 @@
+				MarkdownDescription: "Description of the volume snapshot.",
+			},
+			"performance_policy_id": schema.StringAttribute{
+				Optional:            true,
+				Computed:            true,
+				Description:         "Performance Policy id of the volume snapshot.",
+				MarkdownDescription: "Performance Policy id of the volume snapshot.",
+				Validators: []validator.String{
+					stringvalidator.OneOf([]string{
+						"default_medium",
+						"default_low",
+						"default_high",
+					}...),
+				},
+			},
+			"expiration_timestamp": schema.StringAttribute{
+				Required:            true,
+				Description:         "Expiration Timestamp of the volume snapshot.",
+				MarkdownDescription: "Expiration Timestamp of the volume snapshot.",
+				Validators: []validator.String{
+					stringvalidator.RegexMatches(
+						regexp.MustCompile(`\b[0-9]{4}-[0-9]{2}-[0-9]{2}T[0-9]{2}:[0-9]{2}:[0-9]{2}Z\b`),
+						"Only UTC (+Z) format is allowed eg: 2023-05-06T09:01:47Z",
+					),
+				},
+			},
+			"creator_type": schema.StringAttribute{
+				Optional:            true,
+				Computed:            true,
+				Description:         "Creator Type of the volume snapshot.",
+				MarkdownDescription: "Creator Type of the volume snapshot.",
+				Validators: []validator.String{
+					stringvalidator.OneOf([]string{
+						"User",
+						"System",
+						"Scheduler",
+					}...),
+				},
+			},
+		},
+	}
+}
+
+// Configure - defines configuration for volume snapshot resource
+func (r *resourceSnapshot) Configure(ctx context.Context, req resource.ConfigureRequest, resp *resource.ConfigureResponse) {
+	// Prevent panic if the provider has not been configured.
+	if req.ProviderData == nil {
+		return
+	}
+
+	client, ok := req.ProviderData.(*client.Client)
+
+	if !ok {
+		resp.Diagnostics.AddError(
+			"Unexpected Resource Configure Type",
+			fmt.Sprintf("Expected *http.Client, got: %T. Please report this issue to the provider developers.", req.ProviderData),
+		)
+
+		return
+	}
+
+	r.client = client
+}
+
+// Create - method to create volume snapshot resource
+func (r *resourceSnapshot) Create(ctx context.Context, req resource.CreateRequest, resp *resource.CreateResponse) {
+
+	var plan models.Snapshot
+
+	diags := req.Plan.Get(ctx, &plan)
+	resp.Diagnostics.Append(diags...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	volID := plan.VolumeID.ValueString()
+
+	name := plan.Name.ValueString()
+	description := plan.Description.ValueString()
+	performancePolicyID := plan.PerformancePolicyID.ValueString()
+	expirationTimestamp := plan.ExpirationTimestamp.ValueString()
+	creatorType := plan.CreatorType.ValueString()
+
+	// if volume name is present instead of ID
+	if volID == "" {
+		// Check if both volume name and volume ID are absent
+		if plan.VolumeName.ValueString() == "" {
+			resp.Diagnostics.AddError(
+				"Error creating volume snapshot",
+				"At least one of, volume name OR volume ID should be present",
```

Comment:
Better to add a validation on the schema itself which is "ExactlyOneOf"

Reply:
A validation of conflicts in Schema is already present.

Reply:
I believe instead of conflicts , we can use exactlyOneOf validation .
Reference : https://github.com/dell/terraform-provider-powerflex/blob/main/powerflex/snapshot_resource_schema.go#L37

Reply:
Updated